R/exclude-pregnancy.R

```diff
@@ -0,0 +1,126 @@
+#'     "get_only_insulin_purchases()",
+#'     "get_insulin_purchases_within_180_days()", and
+#'     "get_insulin_is_two_thirds_of_gld_doses()" helper functions for the
+#'     classification of diabetes type.
+#'
+#' @param excluded_pcos Output from [exclude_potential_pcos()].
+#' @param pregnancy_dates Output from [get_pregnancy_dates()].
+#' @param included_hba1c Output from [include_hba1c()].
+#'
+#' @returns The same type as the input data, default as a [tibble::tibble()].
+#'    Has the same output data as the input [exclude_potential_pcos()], except
+#'    for a helper logical variable `no_pregnancy` that is used in later
+#'    functions.
+#' @keywords internal
+#' @inherit algorithm seealso
+#'
+#' @examples
+#' \dontrun{
+#' register_data <- simulate_registers(
+#'   c(
+#'     "lpr_adm",
+#'     "lpr_diag",
+#'     "kontakter",
+#'     "diagnoser",
+#'     "lmdb",
+#'     "bef",
+#'     "lab_forsker"
+#'   ),
+#'   n = 1000
+#' )
+#' lpr2 <- prepare_lpr2(register_data$lpr_adm, register_data$lpr_diag)
+#' lpr3 <- prepare_lpr3(register_data$diagnoser, register_data$kontakter)
+#'
+#' # Exclude pregnancy dates
+#' register_data$lmdb |>
+#'   include_gld_purchases() |>
+#'   exclude_potential_pcos(register_data$bef) |>
+#'   exclude_pregnancy(
+#'     get_pregnancy_dates(lpr2, lpr3),
+#'     include_hba1c(register_data$lab_forsker)
+#'   )
+#' }
+exclude_pregnancy <- function(
+  excluded_pcos,
+  pregnancy_dates,
+  included_hba1c
+) {
+  criteria <- get_algorithm_logic("is_not_within_pregnancy_period") |>
+    # To convert the string into an R expression.
+    rlang::parse_expr()
+
+  # TODO: This should be done at an earlier stage.
+  # Ensure both date columns are of type Date.
+  excluded_pcos <- excluded_pcos |>
+    dplyr::mutate(
+      date = lubridate::as_date(.data$date)
+    )
+  included_hba1c <- included_hba1c |>
+    dplyr::mutate(
+      date = lubridate::as_date(.data$date)
+    )
+
+  excluded_pcos |>
+    dplyr::full_join(
+      included_hba1c,
+      by = dplyr::join_by("pnr", "date")
+    ) |>
+    dplyr::full_join(
+      pregnancy_dates,
+      by = dplyr::join_by("pnr"),
+      relationship = "many-to-many"
+    ) |>
+    # Apply the criteria to flag rows that are within the pregnancy period.
+    dplyr::mutate(
+      is_not_within_pregnancy_period = !!criteria
+    ) |>
+    # Group by pnr and date to ensure the row is dropped if it falls within
+    # *any* pregnancy period. This prevents mistakenly keeping a row just
+    # because it falls outside one pregnancy window, when it may still fall
+    # inside another for the same pnr.
+    dplyr::group_by(.data$pnr, .data$date) |>
+    dplyr::mutate(
+      is_not_within_any_pregnancy_period = all(
+        .data$is_not_within_pregnancy_period == TRUE
+      )
+    ) |>
+    dplyr::ungroup() |>
+    # Only keep rows that don't fall within any pregnancy period.
+    dplyr::filter(.data$is_not_within_any_pregnancy_period) |>
```

Comment:
```suggestion
    # Only keep rows that don't fall within any pregnancy period.
    dplyr::filter(all(.data$is_not_within_pregnancy_period)) |>
    dplyr::ungroup() |>
```

Don't need to make a column just to filter by it :relaxed: